### ISSUE TITLE
Unable to run `nix-shell .` in dhall/ on master

### ISSUE THREAD

**Issue by philandstuff** [dhall-lang/dhall-haskell#1103]
I'm trying to build dhall but I consistently get problems running `nix-shell`.  I have updated submodules and installed the dhall-lang.org binary caches, but I seem to be getting problems downloading `/nix/store/hfi080vzqqzfip6bd6x4cxc2jgj56xn3-ghc-8.4.3`.  Specifically, I frequently see errors of the sort:

    warning: unable to download 'https://cache.dhall-lang.org/nar/hfi080vzqqzfip6bd6x4cxc2jgj56xn3.nar': HTTP error 200 (curl error: Timeout was reached); retrying in 255 ms

followed by some other error (I have seen `non-zero padding` or `std::bad_alloc` afterward).

Has anyone else seen this problem?  I've been able to download lots of other dependencies from cache.dhall-lang.org so it's weird that this one is a problem.

**Comment by Gabriel439** [comment]
@philandstuff: `ghc-8.4.3` is a large dependency (~1.3 GB):

```
$ curl https://cache.dhall-lang.org/hfi080vzqqzfip6bd6x4cxc2jgj56xn3.narinfo
StorePath: /nix/store/hfi080vzqqzfip6bd6x4cxc2jgj56xn3-ghc-8.4.3
URL: nar/hfi080vzqqzfip6bd6x4cxc2jgj56xn3.nar
Compression: none
NarHash: sha256:1hxys3zd1zjdz7rn2i7wg5x7xjmcgs926lv7fzbjdhn8vbb04zql
NarSize: 1354853896
References: 1dn3b6vq7gn7w6kynxy0w1q1vd3ayg80-gcc-wrapper-7.3.0 1xwbnc903gymfi99zpvj8mdp5601yqz6-ncurses-6.1 254rgj92dhkyg7sydb7al65gflnyjd7b-gmp-6.1.2 3a74d03pq6d0cg3crkfw965k5s58hbn4-binutils-2.30 89lj7wjp81pbg4wv2yf8fwm4p5dzq27b-musl-iconv-1.1.19 937b1ps5v3ixg20s7wird5c5szlmw8pa-musl-1.1.19 h0776npra2dhhrm4x9n1z72876p1qmsc-bash-4.4-p23 hfi080vzqqzfip6bd6x4cxc2jgj56xn3-ghc-8.4.3 jgd175yqgpra6g3pcnpvj9cs450722ak-coreutils-8.29 pgsjpn6hp8cgcbrvvabk6vnza2i7l7sg-binutils-wrapper-2.30 q0g2pp86wqlk8diyridand3hifk18z51-ghc-8.4.3-doc rd8w17sgz9wr47nd0s7xvarapib0h9jm-gcc-7.3.0 w3aj7ihnvnfyy909w1z18knb1r2d6afi-perl-5.24.3 wbgg70ij5vjidlj93z07x9bcrgl0i78a-gmp-6.1.2-dev
Deriver: gy2j8b1g5q24q6jzb8kqwqyvxwql0sq9-ghc-8.4.3.drv
Sig: cache.dhall-lang.org:heoXCEj+BFaGWJE9IDx8zsRBQbd4U+QMEsdWXUl6um3DDGKj6a+8RPFKIB+0ASfZT0FWrgp9OEa0ro5vwTPpBg==
```

... so I'm guessing you're hitting some timeout due to the size of the download.

The first thing I would try is to download the same path outside of Nix, using `curl`:

```bash
$ curl -O https://cache.dhall-lang.org/nar/hfi080vzqqzfip6bd6x4cxc2jgj56xn3.nar
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 1292M    0 1292M    0     0  21.6M      0 --:--:--  0:00:59 --:--:-- 21.4M
```

If that fails then either `curl` is bugged or your networking stack is timing out downloads or your ISP is interfering with large downloads.

If that succeeds, then the next step is to test the same download using Nix, since Nix may be using a different `curl` than the `curl` you have installed on your `PATH`:

```bash
$ nix-store --realise /nix/store/hfi080vzqqzfip6bd6x4cxc2jgj56xn3-ghc-8.4.3
```

That will attempt to fetch the path from any available binary substitutes.  If it still fails, you can increase the verbosity by adding something in between `-v` and `-vvvvv` to see if that illuminates what is going wrong at the time of failure.

**Comment by philandstuff** [comment]
:thinking: I'm currently away from home and on the local internet, it has taken me 60 seconds to download 7379k.  I'm hoping my home broadband will be better but for now I think I'll just leave it and open a separate issue for the real problem I was trying to solve.

**Comment by philandstuff** [comment]
I upgraded my internet to 60Mbps and I’ve finally managed to download this dependency, so I’ll close this issue as it Works For Me. However I don’t think it’s a great solution to expect users to have internet this fast. 

I noticed that the server doesn’t output Content-length headers so curl can’t offer a meaningful progress bar, nor can it tell if it has the whole archive or not if the connection is randomly closed (which would explain some of the errors I saw). Might it be possible to at least add Content-length to the binary cache headers?
